ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -357,39 +405,84 @@ func getReplicaIndex(clusterName string, groupName string, namespace string) int
 }
 
 // returns next lowest TPU_WORKER_ID in pod slice and updates mappings
-func getNextWorkerID(podSlice slice, namespace string, replicaIndex int) int {
-	tpuWorkerID := 0
-	if sliceToWorkers[podSlice] == nil {
-		newWorker := worker{tpuWorkerID, replicaIndex, true}
-		sliceToWorkers[podSlice] = []worker{newWorker}
-	} else {
-		nextLowestID := math.MaxInt32
-		replacePod := false
-		// iterate through existing workers and check if any have been deleted
-		for _, worker := range sliceToWorkers[podSlice] {
-			if worker.isCreated == false && worker.workerIndex < nextLowestID {
-				replacePod = true
-				nextLowestID = worker.workerIndex
+func getNextWorkerID(sliceToWorkerIDs map[slice][]int, podSlice slice, namespace string, replicaIndex int) int {
+	tpuWorkerID := 0 // defaults to 0 (first Pod in slice)
+	if sliceToWorkerIDs[podSlice] != nil {
+		sort.Ints(sliceToWorkerIDs[podSlice])
+		// iterate through existing workers and get the next lowest, unused ID
+		for _, workerID := range sliceToWorkerIDs[podSlice] {
+			if workerID == tpuWorkerID {
+				tpuWorkerID++
 			}
 		}
-		// reassign next lowest TPU_WORKER_ID if pod has been deleted
-		if replacePod == true {
-			for index, worker := range sliceToWorkers[podSlice] {
-				// set worker.isCreated to true now that pod is being re-created
-				if worker.workerIndex == nextLowestID {
-					sliceToWorkers[podSlice][index].isCreated = true
+	}
+	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "TPU_WORKER_ID", tpuWorkerID)
+	return tpuWorkerID
+}
+
+// builds mapping representing the current RayCluster state of TPU pods using PodInformer
+func getSliceToWorkerIDs(podLister listersv1.PodLister, clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
```

Comment:
Is it possible to update getSliceToWorkerIDs to receive a list of Pods instead of a podLister? And  in `mutatePods` fetch the pod list using the lister before calling `getSliceToWorkerIDs`. This way you can unit test it separately by passing it a list of Pod objects.

Reply:
Done in [dce1bbe](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723/commits/dce1bbe4b09fd0a48304cb2d519414578a3ecdc0).